htlcswitch/link.go

```diff
@@ -1150,6 +1228,132 @@ func (l *channelLink) htlcManager() {
 				"PendingLocalUpdateCount")
 		}
 
+		// If we are attempting to shutdown the link to cooperatively
+		// close and the channel is clear of our pending updates and we
+		// haven't already sent a Shutdown, we'll send one now.
+		if (l.shutdownInit || l.shutdownReceived) && !l.shutdownSent &&
+			l.channel.PendingLocalUpdateCount() == 0 {
+
+			// Create the Shutdown we'll send over. Send the local
+			// upfront shutdown script if it exists, and the
+			// provided delivery address otherwise. This is used in
+			// case we did not initiate the coop close.
+			upfrontAddr := l.channel.LocalUpfrontShutdownScript()
+			if len(upfrontAddr) != 0 {
+				l.cfg.DeliveryAddr = upfrontAddr
+			} else {
+				// Persist the delivery script for
+				// retransmission.
+				err := l.channel.State().PersistDeliveryScript(
+					l.cfg.DeliveryAddr,
+				)
+				if err != nil {
+					l.fail(LinkFailureError{
+						code: ErrInternalError,
+					}, "failed persisting script: %v", err)
+					return
+				}
+			}
+
+			// Mark that we've sent the Shutdown message. This is
+			// used for retransmission.
+			err := l.channel.State().MarkShutdownSent()
+			if err != nil {
+				l.fail(LinkFailureError{
+					code: ErrInternalError,
+				}, "failed marking shutdown: %v", err)
+				return
+			}
+
+			// Notify peer.Brontide that we've sent the Shutdown
+			// message so that the coop close state machine can
+			// advance.
+			closeReq := l.localCloseReq
+			if closeReq == nil {
+				// If we are not the initiator of the coop
+				// close, we'll create a ChanClose struct. We
+				// won't listen on the error or update chans.
+				// If there's an error, we'll eventually stop
+				// the link anyways. The Updates chan is
+				// currently unused, but is added to prevent a
+				// future change from hanging if it attempts to
+				// send along the chan.
+				closeType := contractcourt.CloseRegular
+				chanPoint := l.ChannelPoint()
+				deliveryAddr := l.cfg.DeliveryAddr
+				errChan := make(chan error, 1)
+				updatesChan := make(chan interface{}, 2)
+
+				// Set the TargetFeePerKw to ensure that if a
+				// bug occurs and this struct is used, that the
+				// zero fee isn't used instead.
+				feeEst := l.cfg.FeeEstimator
+				feePerKw, err := feeEst.EstimateFeePerKW(6)
```

Comment:
remember there's something like a fee floor you could use here.

Reply:
i think `EstimateFeePerKw` will automatically check + enforce a minimum relay fee